# 项目：评估和清理泰坦尼克号幸存者数据

## 分析目标

此数据分析的目的是，根据泰坦尼克号乘客幸存者数据，对幸存者影响因素进行推测。

本实战项目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

> 泰坦尼克号（英语：RMS Titanic）是一艘奥林匹克级邮轮，于1912年4月首航时撞上冰山后沉没。泰坦尼克号是同级的3艘超级邮轮中的第2艘，与姐妹船奥林匹克号和不列颠号为白星航运公司的乘客们提供大西洋旅行。

> 泰坦尼克号由位于北爱尔兰贝尔法斯特的哈兰·沃尔夫船厂兴建，是当时最大的客运轮船，由于其规模相当一艘现代航空母舰，因而号称“上帝也沉没不了的巨型邮轮”。在泰坦尼克号的首航中，从英国南安普敦出发，途经法国瑟堡-奥克特维尔以及爱尔兰昆士敦，计划横渡大西洋前往美国纽约市。但因为人为错误，于1912年4月14日船上时间夜里11点40分撞上冰山；2小时40分钟后，即4月15日凌晨02点20分，船裂成两半后沉入大西洋，死亡人数超越1500人，堪称20世纪最大的海难事件，同时也是最广为人知的海难之一。

数据表：`titianic.csv`。

`titianic.csv`记录了超过八百位泰坦尼克号乘客在沉船事件后的幸存情况，以及乘客的相关信息，包括所在船舱等级、性别、年龄、同乘伴侣/同胞数量、同乘父母/孩子数量，等等。

`titianic.csv`每列的含义如下：
- PassengerId：乘客ID
- survival：是否幸存
   - 0	否
   - 1	是
- pclass：船舱等级
   - 1	一等舱
   - 2	二等舱
   - 3  三等舱
- sex：性别
- Age：年龄
- sibsp：同乘伴侣/同胞数量
- parch：同乘父母/孩子数量
- ticket：船票号
- fare：票价金额
- cabin：船舱号
- embarked：登船港口
   - C  瑟堡
   - Q  皇后镇
   - S  南安普敦

## 读取数据

导入数据分析所需要的库。

In [1]:
import numpy as np
import pandas as pd

通过Pandas的`read_csv`函数，将原始数据文件`titanic.csv`里的数据内容，解析为DataFrame并赋值给变量`original_titanic`。

In [2]:
original_titanic = pd.read_csv("titanic.csv")
original_titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

## 评估和清理数据

在这一部分中，我们将对在上一部分建立的`original_titanic`DataFrame所包含的数据进行评估和清理。

主要从两个方面进行：结构和内容，即整齐度和干净度。

数据的结构性问题指不符合“每个变量为一列，每个观察值为一行，每种类型的观察单位为一个表格”这三个标准；数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

为了区分开经过清理的数据和原始的数据，我们创建新的变量`cleaned_titanic`，让它为`original_titanic`复制出的副本。我们之后的清理步骤都将被运用在`cleaned_titanic`上。

In [3]:
cleaned_titanic = original_titanic.copy()

### 数据整齐度

In [4]:
cleaned_titanic.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

从头部的10行数据来看，数据符合“每个变量为一列，每个观察值为一行，每种类型的观察单位为一个表格”，因此不存在结构性问题。

### 数据干净度

接下来通过`info`，对数据内容进行大致了解。

In [5]:
cleaned_titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


从输出结果来看，`cleaned_titanic`共有891条观察值，其中`Age`、`Cabin`和`Embarked`存在缺失值，将在后续进行评估和清理。

数据类型方面，`PassengerId`表示乘客ID，数据类型不应为数字，应为字符串，所以需要进行数据格式转换。

并且，我们已知`Survived`（是否幸存）、`Pclass`（船舱等级）、`Sex`（性别）、`Embarked`（登船港口）都是分类数据，可以把数据类型都转换为Category。

In [6]:
cleaned_titanic['PassengerId'] = cleaned_titanic['PassengerId'].astype('str')

In [7]:
cleaned_titanic['Survived'] = cleaned_titanic['Survived'].astype('category')
cleaned_titanic['Pclass'] = cleaned_titanic['Pclass'].astype('category')
cleaned_titanic['Sex'] = cleaned_titanic['Sex'].astype('category')
cleaned_titanic['Embarked'] = cleaned_titanic['Embarked'].astype('category')

In [8]:
cleaned_titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    object  
 1   Survived     891 non-null    category
 2   Pclass       891 non-null    category
 3   Name         891 non-null    object  
 4   Sex          891 non-null    category
 5   Age          714 non-null    float64 
 6   SibSp        891 non-null    int64   
 7   Parch        891 non-null    int64   
 8   Ticket       891 non-null    object  
 9   Fare         891 non-null    float64 
 10  Cabin        204 non-null    object  
 11  Embarked     889 non-null    category
dtypes: category(4), float64(2), int64(2), object(4)
memory usage: 59.8+ KB


#### 处理缺失数据

从`info`方法的输出结果来看，在`cleaned_titanic`中，`Age`、`Cabin`和`Embarked`变量存在缺失值。

In [9]:
cleaned_titanic[cleaned_titanic['Age'].isna()]

PassengerId Survived Pclass                                      Name  \
5             6        0      3                          Moran, Mr. James   
17           18        1      2              Williams, Mr. Charles Eugene   
19           20        1      3                   Masselmani, Mrs. Fatima   
26           27        0      3                   Emir, Mr. Farred Chehab   
28           29        1      3             O'Dwyer, Miss. Ellen "Nellie"   
..          ...      ...    ...                                       ...   
859         860        0      3                          Razi, Mr. Raihed   
863         864        0      3         Sage, Miss. Dorothy Edith "Dolly"   
868         869        0      3               van Melkebeke, Mr. Philemon   
878         879        0      3                        Laleff, Mr. Kristo   
888         889        0      3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
5      male  NaN      0      0      330877   8.4583   NaN        Q  
17     male  NaN      0      0      244373  13.0000   NaN        S  
19   female  NaN      0      0        2649   7.2250   NaN        C  
26     male  NaN      0      0        2631   7.2250   NaN        C  
28   female  NaN      0      0      330959   7.8792   NaN        Q  
..      ...  ...    ...    ...         ...      ...   ...      ...  
859    male  NaN      0      0        2629   7.2292   NaN        C  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S  
868    male  NaN      0      0      345777   9.5000   NaN        S  
878    male  NaN      0      0      349217   7.8958   NaN        S  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S  

[177 rows x 12 columns]

有177条观察值的年龄变量缺失，占总体数据比例20%左右。由于这些观察值数量较多，且的其它变量仍然能为分析提供价值，我们最好保留这些行，用乘客年龄平均值对缺失值进行填充。

In [10]:
average_age = cleaned_titanic['Age'].mean()
cleaned_titanic['Age'] = cleaned_titanic['Age'].fillna(average_age)
cleaned_titanic['Age'].isna().sum()

np.int64(0)

In [11]:
cleaned_titanic[cleaned_titanic['Cabin'].isna()]

PassengerId Survived Pclass                                      Name  \
0             1        0      3                   Braund, Mr. Owen Harris   
2             3        1      3                    Heikkinen, Miss. Laina   
4             5        0      3                  Allen, Mr. William Henry   
5             6        0      3                          Moran, Mr. James   
7             8        0      3            Palsson, Master. Gosta Leonard   
..          ...      ...    ...                                       ...   
884         885        0      3                    Sutehall, Mr. Henry Jr   
885         886        0      3      Rice, Mrs. William (Margaret Norton)   
886         887        0      2                     Montvila, Rev. Juozas   
888         889        0      3  Johnston, Miss. Catherine Helen "Carrie"   
890         891        0      3                       Dooley, Mr. Patrick   

        Sex        Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0      male  22.000000      1      0         A/5 21171   7.2500   NaN        S  
2    female  26.000000      0      0  STON/O2. 3101282   7.9250   NaN        S  
4      male  35.000000      0      0            373450   8.0500   NaN        S  
5      male  29.699118      0      0            330877   8.4583   NaN        Q  
7      male   2.000000      3      1            349909  21.0750   NaN        S  
..      ...        ...    ...    ...               ...      ...   ...      ...  
884    male  25.000000      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.000000      0      5            382652  29.1250   NaN        Q  
886    male  27.000000      0      0            211536  13.0000   NaN        S  
888  female  29.699118      1      2        W./C. 6607  23.4500   NaN        S  
890    male  32.000000      0      0            370376   7.7500   NaN        Q  

[687 rows x 12 columns]

有687条观察值的船舱号变量缺失，说明船舱号数据在大部分观察值中都是未知的，所以不能删除这些观察值。

In [12]:
cleaned_titanic[cleaned_titanic['Embarked'].isna()]

PassengerId Survived Pclass                                       Name  \
61           62        1      1                        Icard, Miss. Amelie   
829         830        1      1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
61   female  38.0      0      0  113572  80.0   B28      NaN  
829  female  62.0      0      0  113572  80.0   B28      NaN

仅有两条观察值的登船港口变量缺失，但我们认为登船港口并不是影响生还概率的关键因素，因此可以保留这些观察值。

#### 处理重复数据

根据数据变量的含义以及内容来看，`PassengerId`是乘客的唯一标识符，不应该存在重复，因此查看是否存在重复值。

In [13]:
cleaned_titanic["PassengerId"].duplicated().sum()

np.int64(0)

输出结果为0，说明不存在重复值。

#### 处理不一致数据

不一致数据可能存在于所有分类变量中，我们要查看是否存在不同值实际指代同一目标的情况。

In [14]:
cleaned_titanic["Survived"].value_counts()

Survived
0    549
1    342
Name: count, dtype: int64

In [15]:
cleaned_titanic["Pclass"].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [16]:
cleaned_titanic["Sex"].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

In [17]:
cleaned_titanic["Embarked"].value_counts()

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

从以上输出结果来看，均不存在不一致数据。

#### 处理无效或错误数据

可以通过DataFrame的`describe`方法，对数值统计信息进行快速了解。

In [18]:
cleaned_titanic.describe()

Age       SibSp       Parch        Fare
count  891.000000  891.000000  891.000000  891.000000
mean    29.699118    0.523008    0.381594   32.204208
std     13.002015    1.102743    0.806057   49.693429
min      0.420000    0.000000    0.000000    0.000000
25%     22.000000    0.000000    0.000000    7.910400
50%     29.699118    0.000000    0.000000   14.454200
75%     35.000000    1.000000    0.000000   31.000000
max     80.000000    8.000000    6.000000  512.329200

乘客年龄平均为30岁左右，最大值为80岁，最小值为0.42岁。同乘伴侣/同胞数量最大值为8个，最小为0个。同乘父母/孩子数量最大值为6个，最小值为0个。船票价格平均为32元，最大值为512元，最小值为0元，猜测0元表示增票。数据不存在脱离现实的数值。